### DISCUSSION TITLE
Adding support for registration of Segment Routing

### DISCUSSION THREAD

**Discussion by rinsekloek** [nautobot/nautobot#1014] [Plugin (App) Ideas & Development]
### Environment
* Python version: 3.7
* Nautobot version: v1.0.3

### Proposed Functionality
Currently we see more and more ISP networks roll out Segment Routing (mostly SR-MPLS). In segment routing MPLS it is essential that there is a Global SRGB in which every Segment (router+loopback) has a uniqe SID .

We would like to see that Nautobot has a Model to have a registration of these SID's that are used in a Segment Routing network( SRGB). A SID is commonly (in ISIS) linked to an interface (Loopback for example):

Router1:
router isis CORE
 interface Loopback0
  passive
  address-family ipv4 unicast
   prefix-sid index 4000 
  !
  address-family ipv6 unicast
   prefix-sid index 6000 
  !
 !
Router2:
router isis CORE
 interface Loopback0
  passive
  address-family ipv4 unicast
   prefix-sid index 4001 
  !
  address-family ipv6 unicast
   prefix-sid index 6001
  !
 !

See https://datatracker.ietf.org/doc/html/rfc8402

### Use Case

Use case would be every Network that deploys Segment Routing would have a central databas where the SID's can be registered. This will overcome the situation of having duplicate SID's in the networking resulting in routing errors. (See it like duplicate loopback IP addresses)

### Database Changes

For the implementation of Segment Routing registration, we will need two extra table with some fields that can be linked to an interface.

The table would have fields like:
1 SR Domains (Like VLAN Groups)
SR domain name | Description | (optional fields like tenant /site)
2 Segment Routing DID's
SR Domain |  SID . The combination SID+SR Domain needs to be unique.

**Comment by glennmatthews** [comment]
Thanks for the request! This would be very much worth initially proofing as a plugin. For the two relatively simple data models you've described, a plugin should not be difficult to implement and explore. If you don't have experience [writing plugins](https://nautobot.readthedocs.io/en/stable/plugins/development/), the documentation can help, and joining the `#nautobot` channel on slack.networktocode.com would be a great place to start asking questions.

For the time being, I'm going to convert this to a discussion thread where the specific requirements of a more broadly capable segment-routing data model in Nautobot can be worked out.

**Reply by josh9730** [reply]
I've been working on a plugin for this: [nautobot_app_sid](https://github.com/josh9730/refactored-couscous/tree/nautobot/apps/nautobot_app_sid). Mostly complete as-is except for some tests, and I'd like to add a Choice field for the Node SID selection (based on available SIDs) rather than the current integer field, but not quite sure how to go about that yet.

I ignored the SRGB entirely, as my use-case has the same SRGB across a single SR domain. I also did not add tenant/site/etc as these would be under the device itself. This also makes some assumptions that may not be valid for your environment (such as IPv6 SID being IPv4 SID + 1) but should be straightforward to adjust to meet your needs.

**Reply by hvisage** [reply]
Is the nautobot app for the SID still working/worked on?
